### ISSUE TITLE
listenerApi.condition not working as expected

### ISSUE THREAD

**Issue by skavan** [reduxjs/redux-toolkit#2698]
Hi,

Forgive any poor terminology, I'm relatively new at redux.

I have created a test reducer (`uselessReducer`), a `listenerMiddleware`, and my general `appMiddleware`.
`uselessReducer` has an action, `updateStatus`. the `listenerMiddleware `has an `actionCreator `on `updateStatus`.

In my general `appMiddleware`, if the useless/init case is triggered, the `updateStatus` action is dispatched and my listener is triggered, its `condition` is met (connected: true) and it executes as expected. All good.

```    
    return function (next) {
        return function (action) {
            switch (action.type) {
                case "useless/init":
                    store.dispatch(updateStatus({ connected: true, status: "connected" }));
                    return
                default:
                    break;
            }
            return next(action);
        };
    };
```

Here's the listener:
```
listenerMiddleware.startListening({
  actionCreator: updateStatus,
  effect: async (action, listenerApi) => {
        // Can cancel other running instances
        listenerApi.cancelActiveListeners()  
        console.log("LISTENER RUNNING")

        if (await listenerApi.condition((action, currentState) =>
          currentState.useless.connected === true
        )){
          console.log("LISTENER TRIGGERED")
        }
        
  }
})
```

However, if I change the appMiddleware and dispatch the action inside the .then(..), then the action is correctly dispatched, the listener is run, but the condition doesn't work and the following code is never executed. I must be missing something pretty fundamental, but can't figure out what! What am I missing here?

```
case "useless/init":
       axios.get("http://somewebsite.xxx/").then((res) => {
               store.dispatch(updateStatus({ connected: true, status: "connected" }));                    
       });
       return
```

**Comment by markerikson** [comment]
Can you post a CodeSandbox that shows the actual code, and possibly a Replay ( https://replay.io/record-bugs ) that shows this happening?  I don't see enough to understand what the setup is here.

**Comment by phryneas** [comment]
The problem is that that `await listenerApi.condition(` will test on every future action dispatch if that condition is true - but you are not dispatching any more actions.

It's a bit awkward, because at the point in time where you do the `await condition`, the condition is actually already true, so it feels like it should execute immediately. But of course that is problematic as well, because `condition` does not only test for state, but also for actions - so it would need an action. Since there is no action, it cannot run immediately.

I think we might want to still execute that `condition` immediately anyways - with some fake action like `{ type: 'RTK__immediateConditionCheck' }`. @markerikson what do you think?

**Comment by markerikson** [comment]
I don't think I understand what the intent is here.

The point of `await condition()` _is_ to wait and check after _future_ actions, not "right now".

If you want to check what the state is _now_, you should use `getState()` and compare some value inside.

Can you clarify what you're trying to accomplish?

**Comment by phryneas** [comment]
I'd assume it's a "wait for this state to be the case. It might already be the case or happen in the future".

I guess a
```js
if (getState().someCondition || await listenerApi.condition((action, state) => state.someCondition) {
  ...
}
```
just feels like unneccessary repetition for something like that - so I get the wish for this to work.

**Comment by skavan** [comment]
> I'd assume it's a "wait for this state to be the case. It might already be the case or happen in the future". 

@phryneas -- exactly correct!

**Comment by markerikson** [comment]
Yeah, the listener middleware's implementation is explicitly "run checks _after_ each action is dispatched", and `condition()` is implemented internally as yet another listener instance.  So, it _won't_ check right away - you would have to do dispatch some action to trigger an initial check, or do the check manually yourself.

I do get that's not ideal, but we'd have to do some major rethinking of how this works otherwise.

**Comment by skavan** [comment]
fair enough.  the conditional approach above will work...

**Comment by phryneas** [comment]
> I do get that's not ideal, but we'd have to do some major rethinking of how this works otherwise.

See my suggestion above - essentially
```diff
             condition: (
               predicate: AnyListenerPredicate<any>,
               timeout?: number
-            ) => take(predicate, timeout).then(Boolean),
+            ) =>
+              predicate(
+                { type: 'RTK__immediateConditionCheck' },
+                api.getState(),
+                getOriginalState()
+              )
+                ? Promise.resolve(true)
+                : take(predicate, timeout).then(Boolean),
             take,
```

**Comment by skavan** [comment]
But -- does that force an immediate ONLY check (or no?). Your original form, was an either/or which is what we want the outcome to be.

**Comment by phryneas** [comment]
It's a suggested change in https://github.com/reduxjs/redux-toolkit/blob/a60e79d9dd87f393f16d991d6fab9623ef719954/packages/toolkit/src/listenerMiddleware/index.ts#L385-L388 - which would immediately do the check and if that fails go back to the current behaviour.

**Comment by skavan** [comment]
By the way -- I dunno if this was implied -- by 
> you would have to do dispatch some action to trigger an initial check
But what I would really like to do is init a listener, without triggering a dispatch from the app side of things. i.e a kind of autorun. I'm too inexperienced to figure out how to do that!

**Comment by skavan** [comment]
> - which would immediately do the check and if that fails go back to the current behaviour.

perfect. but if its too much brain damage, I'm ok with as its pretty self explanatory.

`if (getState().someCondition || await listenerApi.condition((action, state) => state.someCondition) {
  ...
}`

**Comment by markerikson** [comment]
We _have_ had a couple cases where it would be nice to trigger a listener right away.

One conceptual issue: `predicate` accepts an action. If we trigger it _immediately_, then there isn't an `action` to pass in as the first arg, so what would we do?  If your predicate code happens to check that, and it's `null` instead, it'll explode.  (Alternately, if we were to change the type to be `action: AnyAction | null`, now you have to double-check that in _every_ condition...)

**Comment by skavan** [comment]
> Oh. I see Lenz's point - pass in a dummy action with a fake `type` field, so it at least fulfills the minimum expected contents of an action object.
> 
> That could work, actually...

But wouldn't that still require a triggering event of some action or other in order to hit the routine at all?

**Comment by skavan** [comment]
> We _have_ had a couple cases where it would be nice to trigger a listener right away.

What we kinda want is "when store loaded, run this..."
